### ISSUE TITLE
CSV Filter causes occasional _csvparsefailure when importing CSV files

### ISSUE THREAD

**Issue by Rumbles** [logstash-plugins/logstash-filter-csv#8]
Hi,

I'm using LS 1.5.0 RC2

In my setup I have a script which checks an IMAP account, finds new email, unzips the zip file attached to the email to a folder on the system. I have configured the logstash file input to expand the blob and check every hour, as the script will run at the most daily (it's a python script run from cron). I have written a function in my script to check if the unzipped folders are over 5 days old, if they are, it then removes them.

I check the logs every day, and in the past few days I have noticed a few _csvparsefailures, checking the events it looks like the event is missing some of the first characters in the csv line. Checking the csvs that line came from the line is complete. So I guess either I'm doing something wrong (very possible) or this is a bug?

My input config:

```
input {
  file {
    path => "/var/log/untangle-reports/reports*Daily/reports/*/*.csv"
    type => "untangle-reports"
    start_position => "beginning"
    tags => "drop"
    discover_interval => 3600
  }
}
```

My filter config is over 300 lines long as there are conditionals for each different file name within the zip, each file has a different series of column names, which are defined within the config. I'll add that to a subsequent post.

An example message:

```
{
  "_index": "logstash-2015.05.05",
  "_type": "untangle-reports",
  "_id": "AU0iAsaHl3tTMJhXux8C",
  "_score": null,
  "_source": {
    "message": [
      "port\"\r"
    ],
    "@version": "1",
    "@timestamp": "2015-05-05T02:58:35.604Z",
    "type": "untangle-reports",
    "tags": [
      "mail-events",
      "_csvparsefailure"
    ],
    "host": "lumberjack.domain.com",
    "path": "/var/log/untangle-reports/reports-05_04_15-Daily/reports/untangle-node-virusblocker/mail-events.csv",
    "filename": "mail-events"
  },
  "fields": {
    "@timestamp": [
      1430794715604
    ]
  },
  "highlight": {
    "tags": [
      "@kibana-highlighted-field@_csvparsefailure@/kibana-highlighted-field@"
    ],
    "tags.raw": [
      "@kibana-highlighted-field@_csvparsefailure@/kibana-highlighted-field@"
    ]
  },
  "sort": [
    1430794715604
  ]
}
```

The only line in the file specified is the following:

> "time_stamp","hostname","username","virusblocker_name","subject","addr","c_client_addr","c_client_port"

I am aware there is a plan to improve the csv (e.g. with https://github.com/elastic/logstash/issues/2088 ) input, which would mean a lot of the config I have set up will no longer be required (that is, the part where I define the column names) but at present  don't know another way to bring the data in.

My output config, in case it is relevant:

```
output {
  if "drop" not in [tags] {
    elasticsearch {
      protocol => "http"
      host => localhost
      workers => 4
    }
    #stdout { codec => rubydebug }
  }
}
```

**Comment by Rumbles** [comment]
Here is my filter:

    filter {
      if [type] == "untangle-reports" {
        # Grab the filename from the path and set that as a tag so we know which filter to use
        grok {
          match => { "path" => "%{GREEDYDATA}/%{GREEDYDATA:filename}\.csv" }
        }
        mutate {
          add_tag => [ "%{filename}" ]
        }
        # Filter by tags to apply the right filter
        if "ad-events" in [tags] {
          csv {
            columns => [
              "time_stamp", "client_addr", "login_name", "domain", "type"
            ]
          }
          if [time_stamp] == "time_stamp" {
            drop { }
          }
          mutate {
            remove_tag => "drop"
          }
        }
    
        if "capture_rule" in [tags] {
          csv {
            columns => [
              "session_id", "event_id", "time_stamp", "end_time", "hostname", "username", "policy_id", "c_client_addr", "c_server_addr", "c_server_port", "c_client_port", "s_client_addr", "s_server_addr", "s_server_port", "s_client_port", "client_intf", "server_intf", "c2p_bytes", "p2c_bytes", "s2p_bytes", "p2s_bytes", "shield_blocked", "firewall_blocked", "firewall_flagged", "firewall_rule_index", "protofilter_protocol", "protofilter_blocked", "capture_blocked", "capture_rule_index", "classd_application", "classd_protochain", "classd_blocked", "classd_flagged", "classd_confidence", "classd_ruleid", "classd_detail", "ips_blocked", "ips_ruleid", "ips_description", "bandwidth_priority", "bandwidth_rule", "https_ruleid", "https_status", "https_detail"
            ]
          }
          if [session_id] == "session_id" {
            drop { }
          }
          mutate {
            remove_tag => "drop"
          }
        }
    
        if "capture_user_events" in [tags] {
          csv {
            columns => [
              "time_stamp", "client_addr", "login_name", "event_info"
            ]
          }
          if [time_stamp] == "time_stamp" {
            drop { }
          }
          mutate {
            remove_tag => "drop"
          }
        }
    
        if "detection_events" in [tags] {
          csv {
            columns => [
              "session_id", "event_id", "time_stamp", "end_time", "hostname", "username", "policy_id", "c_client_addr", "c_server_addr", "c_server_port", "c_client_port", "s_client_addr", "s_server_addr", "s_server_port", "s_client_port", "client_intf", "server_intf", "c2p_bytes", "p2c_bytes", "s2p_bytes", "p2s_bytes", "shield_blocked", "firewall_blocked", "firewall_flagged", "firewall_rule_index", "protofilter_protocol", "protofilter_blocked", "capture_blocked", "capture_rule_index", "classd_application", "classd_protochain", "classd_blocked", "classd_flagged", "classd_confidence", "classd_ruleid", "classd_detail", "ips_blocked", "ips_ruleid", "ips_description", "bandwidth_priority", "bandwidth_rule", "https_ruleid", "https_status", "https_detail"
            ]
          }
          if [session_id] == "session_id" {
            drop { }
          }
          mutate {
            remove_tag => "drop"
          }
        }
    
        if "firewall-events" in [tags] {
          csv {
            columns => [
              "session_id", "event_id", "time_stamp", "end_time", "hostname", "username", "policy_id", "c_client_addr", "c_server_addr", "c_server_port", "c_client_port", "s_client_addr", "s_server_addr", "s_server_port", "s_client_port", "client_intf", "server_intf", "c2p_bytes", "p2c_bytes", "s2p_bytes", "p2s_bytes", "shield_blocked", "firewall_blocked", "firewall_flagged", "firewall_rule_index", "protofilter_protocol", "protofilter_blocked", "capture_blocked", "capture_rule_index", "classd_application", "classd_protochain", "classd_blocked", "classd_flagged", "classd_confidence", "classd_ruleid", "classd_detail", "ips_blocked", "ips_ruleid", "ips_description", "bandwidth_priority", "bandwidth_rule", "https_ruleid", "https_status", "https_detail"
            ]
          }
          if [session_id] == "session_id" {
            drop { }
          }
          mutate {
            remove_tag => "drop"
          }  
        }
    
        if "attack-events" in [tags] {
          csv {
            columns => [
              "session_id", "event_id", "time_stamp", "end_time", "hostname", "username", "policy_id", "c_client_addr", "c_server_addr", "c_server_port", "c_client_port", "s_client_addr", "s_server_addr", "s_server_port", "s_client_port", "client_intf", "server_intf", "c2p_bytes", "p2c_bytes", "s2p_bytes", "p2s_bytes", "shield_blocked", "firewall_blocked", "firewall_flagged", "firewall_rule_index", "protofilter_protocol", "protofilter_blocked", "capture_blocked", "capture_rule_index", "classd_application", "classd_protochain", "classd_blocked", "classd_flagged", "classd_confidence", "classd_ruleid", "classd_detail", "ips_blocked", "ips_ruleid", "ips_description", "bandwidth_priority", "bandwidth_rule", "https_ruleid", "https_status", "https_detail"
            ]
          }
          if [session_id] == "session_id" {
            drop { }
          }
          mutate {
            remove_tag => "drop"
          }
        }
    
        if "login_events" in [tags] {
          csv {
            columns => [
              "time_stamp", "client_name", "remote_address", "remote_port"
            ]
          }
          if [time_stamp] == "time_stamp" {
            drop { }
          }
          mutate {
            remove_tag => "drop"
          }
        }
    
        if "phish-events" in [tags] {
          csv {
            columns => [
              "time_stamp", "hostname", "phish_score", "from_addrs", "subject", "s_server_addr", "action", "to_addrs"
             ]
          }
          if [time_stamp] == "time_stamp" {
            drop { }
          }
          mutate {
            remove_tag => "drop"
          }
        }
    
        if "phish-events-all" in [tags] {
          csv {
            columns => [
              "time_stamp", "hostname", "phish_score", "from_addrs", "subject", "s_server_addr", "action", "to_addrs"
            ]
          }
          if [time_stamp] == "time_stamp" {
            drop { }
          }
          mutate {
            remove_tag => "drop"
          }
        }
    
        if "shield-events" in [tags] {
          csv {
            columns => [
              "session_id", "event_id", "time_stamp", "end_time", "hostname", "username", "policy_id", "c_client_addr", "c_server_addr", "c_server_port", "c_client_port", "s_client_addr", "s_server_addr", "s_server_port", "s_client_port", "client_intf", "server_intf", "c2p_bytes", "p2c_bytes", "s2p_bytes", "p2s_bytes", "shield_blocked", "firewall_blocked", "firewall_flagged", "firewall_rule_index", "protofilter_protocol", "protofilter_blocked", "capture_blocked", "capture_rule_index", "classd_application", "classd_protochain", "classd_blocked", "classd_flagged", "classd_confidence", "classd_ruleid", "classd_detail", "ips_blocked", "ips_ruleid", "ips_description", "bandwidth_priority", "bandwidth_rule", "https_ruleid", "https_status", "https_detail"
            ]
          }
          if [session_id] == "session_id" {
            drop { }
          }
          mutate {
            remove_tag => "drop"
          }
        }
        
        if "domains" in [tags] {
          csv {
            columns => [
              "host", "hits", "size"
            ]
          }
          if [host] == "host" {
            drop { }
          }
          mutate {
            remove_tag => "drop"
          }
        }
    
        if "events" in [tags] {
          csv {
            columns => [
              "event_id", "time_stamp", "session_id", "client_intf", "server_intf", "c_client_addr", "s_client_addr", "c_server_addr", "s_server_addr", "c_client_port", "s_client_port", "c_server_port", "s_server_port", "policy_id", "username", "hostname", "request_id", "method", "uri", "host", "c2s_content_length", "s2c_content_length", "s2c_content_type", "adblocker_blocked", "adblocker_cookie_ident", "adblocker_action", "webfilter_reason", "webfilter_category", "webfilter_blocked", "webfilter_flagged", "sitefilter_reason", "sitefilter_category", "sitefilter_blocked", "sitefilter_flagged", "clam_clean", "clam_name", "virusblocker_clean", "virusblocker_name", "url"
            ]
          }
          if [event_id] == "event_id" {
            drop { }
          }
          mutate {
            remove_tag => "drop"
          }
        }
        
        if "queries" in [tags] {
          csv {
            columns => [
              "event_id", "time_stamp", "session_id", "client_intf", "server_intf", "c_client_addr", "s_client_addr", "c_server_addr", "s_server_addr", "c_client_port", "s_client_port", "c_server_port", "s_server_port", "policy_id", "username", "hostname", "request_id", "method", "uri", "term", "host", "c2s_content_length", "s2c_content_length", "s2c_content_type"
            ]
          }
          if [event_id] == "event_id" {
            drop { }
          }
          mutate {
            remove_tag => "drop"
          }
        }
    
        if "unblocks" in [tags] {
          csv {
            columns => [
              "event_id", "time_stamp", "session_id", "client_intf", "server_intf", "c_client_addr", "s_client_addr", "c_server_addr", "s_server_addr", "c_client_port", "s_client_port", "c_server_port", "s_server_port", "policy_id", "username", "hostname", "request_id", "method", "uri", "host", "c2s_content_length", "s2c_content_length", "s2c_content_type", "adblocker_blocked", "adblocker_cookie_ident", "adblocker_action", "webfilter_reason", "webfilter_category", "webfilter_blocked", "webfilter_flagged", "sitefilter_reason", "sitefilter_category", "sitefilter_blocked", "sitefilter_flagged", "clam_clean", "clam_name", "virusblocker_clean", "virusblocker_name"
            ]
          }
          if [event_id] == "event_id" {
            drop { }
          }
          mutate {
            remove_tag => "drop"
          }
        }
    
        if "violations" in [tags] {
          csv {
            columns => [
              "event_id", "time_stamp", "session_id", "client_intf", "server_intf", "c_client_addr", "s_client_addr", "c_server_addr", "s_server_addr", "c_client_port", "s_client_port", "c_server_port", "s_server_port", "policy_id", "username", "hostname", "request_id", "method", "uri", "host", "c2s_content_length", "s2c_content_length", "s2c_content_type", "adblocker_blocked", "adblocker_cookie_ident", "adblocker_action", "webfilter_reason", "webfilter_category", "webfilter_blocked", "webfilter_flagged", "sitefilter_reason", "sitefilter_category", "sitefilter_blocked", "sitefilter_flagged", "clam_clean", "clam_name", "virusblocker_clean", "virusblocker_name", "url"
            ]
          }
          if [event_id] == "event_id" {
            drop { }
          }
          mutate {
            remove_tag => "drop"
          }
        }
    
        if "spam-events" in [tags] {
          csv {
            columns => [
              "time_stamp", "hostname", "spamblocker_score", "from_addrs", "subject", "s_server_addr", "action", "to_addrs"
            ]
          }
          if [time_stamp] == "time_stamp" {
            drop { }
          }
          mutate {
            remove_tag => "drop"
          }
        }
    
        if "spam-events-all" in [tags] {
          csv {
            columns => [
              "time_stamp", "hostname", "spamblocker_score", "from_addrs", "subject", "s_server_addr", "action", "to_addrs"
            ]
          }
          if [time_stamp] == "time_stamp" {
            drop { }
          }
          mutate {
            remove_tag => "drop"
          }
        }
    
        if "ftp-events" in [tags] {
          csv {
            columns => [
              "time_stamp", "hostname", "username", "virusblocker_name", "uri", "s_server_addr"
            ]
          }
          if [time_stamp] == "time_stamp" {
            drop { }
          }
          mutate {
            remove_tag => "drop"
          }
        }
    
        if "mail-events" in [tags] {
          csv {
            columns => [
              "time_stamp", "hostname", "username", "virusblocker_name", "subject", "addr", "c_client_addr", "c_client_port"
            ]
          }
          if [time_stamp] == "time_stamp" {
            drop { }
          }
          mutate {
            remove_tag => "drop"
          }
        }
    
        if "web-events" in [tags] {
          csv {
            columns => [
              "event_id", "time_stamp", "session_id", "client_intf", "server_intf", "c_client_addr", "s_client_addr", "c_server_addr", "s_server_addr", "c_client_port", "s_client_port", "c_server_port", "s_server_port", "policy_id", "username", "hostname", "request_id", "method", "uri", "host", "c2s_content_length", "s2c_content_length", "s2c_content_type", "adblocker_blocked", "adblocker_cookie_ident", "adblocker_action", "webfilter_reason", "webfilter_category", "webfilter_blocked", "webfilter_flagged", "sitefilter_reason", "sitefilter_category", "sitefilter_blocked", "sitefilter_flagged", "clam_clean", "clam_name", "virusblocker_clean", "virusblocker_name", "url"
            ]
          }
          if [event_id] == "event_id" {
            drop { }
          }
          mutate {
            remove_tag => "drop"
          }
        }
    
        if "admin-login-events" in [tags] {
          csv {
            columns => [
              "time_stamp", "client_addr", "login", "succeeded"
            ]
          }
          if [time_stamp] == "time_stamp" {
            drop { }
          }
          mutate {
            remove_tag => "drop"
          }
        }
    
        if "alert-events" in [tags] {
          csv {
            columns => [
              "time_stamp", "description", "summary_text", "json"
            ]
          }
          if [time_stamp] == "time_stamp" {
            drop { }
          }
          mutate {
            remove_tag => "drop"
          }
        }
    
      }
    }

As you can see I don't do any stream editing on the input text, I just have a lot of conditionals. If there is something I am doing wrong here please let me know!

**Comment by renev7** [comment]
@Rumbles did you figured this out? I'm having csv file import issues too. Just like you, looks like import is missing some of the first chars. By looking at the csv source, the entire line looks perfect to me. I'm importing many csv files just once and deleting them before saving new csv files the next day. And the process repeats. First import works fine but, subsequent imports eventually fail.

My logstash config is pretty small:

```
input {
   file {
      path => "/home/financedb/csvdata/*.csv"
      start_position => "beginning"
   }
}

filter {
   csv {
      columns => ["symbol", "date", "open", "high", "low", "close", "volume", "adjclose"]
      separator => ","
   }
}

output {
   elasticsearch {
      action => "index"
      host => "localhost"
      index => "financedb"
      document_type => "close"
      workers => 1
   }
   stdout {
        codec => rubydebug
   }
}
```

**Comment by Rumbles** [comment]
Afraid not, I didn't have time to look in to it, now I work for a different
company.. Sorry

**Comment by purbon** [comment]
@renev7 to reproduce this issue, do you have a sample CSV file that we can use to track this problem down? thanks a lot.

**Comment by renev7** [comment]
@purbon I just moved to a different approach but, give me some time to go back and see if I can reproduce it with a new set.

**Comment by dwasss** [comment]
Just wanted to say I am on LS 7.x and still experience this problem...it seems the source CSV is not parsed properly every time Logstash processes that CSV file and looks like the `message` field is only partially parsed. It keeps happening for a few certain events, those which are the last 3 rows in the CSV, but they look fine to me in the CSV file.

For my `file` input, I do have `start_position => "beginning"`, so I'm not sure if that has something to do with it...

**Comment by kares** [comment]
@dwasss if you have a reporoducer please open a new issue - with a sample content that has valid CSV and fails to parse.
